### PR TITLE
MONGOCRYPT-162: Implement cryptography callbacks for C# libmongocrypt wrapper.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG: https://evergreen.mongodb.com/version/5fadce583627e0079e295dc7

### CODE REVIEW

File: bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs [hunk 1]
```diff
@@ -585,7 +585,7 @@ public void TestLocalKeyCreationStepwise()
         static IEnumerable<string> FindTestDirectories()
         {
             string[] searchPaths = new[] { Path.Combine("..", "test", "example"), Path.Combine("..", "test", "data") };
-            var assemblyLocation = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            var assemblyLocation = Path.GetDirectoryName(typeof(BasicTests).GetTypeInfo().Assembly.Location);
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt.Test/BasicTests.cs:588]: the previous way doesn't work with `netstandard2.0`

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+using FluentAssertions;
+using MongoDB.Libmongocrypt.Callbacks;
+using Xunit;
+
+namespace MongoDB.Libmongocrypt.Test.Callbacks
+{
+    public class CipherCallbacksTests
+    {
+        [Fact]
+        public void CipherTest()
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs:25]: test data can be found here: https://docs.google.com/document/d/1-qmr6miUug2qciOlX2Q7eisChvG_RtzLNJnp6ZtmtvY/edit?ts=5f93353d#heading=h.mp30qfn97460

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/HashCallbackTests.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+using MongoDB.Libmongocrypt.Callbacks;
+using Xunit;
+
+namespace MongoDB.Libmongocrypt.Test.Callbacks
+{
+    public class HashCallbackTests
+    {
+        [Fact]
+        public void HashTest()
+        {
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/HashCallbackTests.cs:26]: test data can be found here https://docs.google.com/document/d/1-qmr6miUug2qciOlX2Q7eisChvG_RtzLNJnp6ZtmtvY/edit?ts=5f93353d#heading=h.mp30qfn97460

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/HmacShaCallbacksTests.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+using FluentAssertions;
+using MongoDB.Libmongocrypt.Callbacks;
+using Xunit;
+
+namespace MongoDB.Libmongocrypt.Test.Callbacks
+{
+    public class HmacShaCallbacksTests
+    {
+        [Theory]
+        [InlineData(256, "74657374206f66206d6163", "37626663386235656333306537336439386565666133653263633334643635376535323734623537656633326661353862663638313534396535663737303138", "ebfaa874ff7fcf5b48637a4aff49ed60f48b53a0802719d6ad85f96d315b2df2")]
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/HmacShaCallbacksTests.cs:25]: test data can be found here https://docs.google.com/document/d/1-qmr6miUug2qciOlX2Q7eisChvG_RtzLNJnp6ZtmtvY/edit?ts=5f93353d#heading=h.mp30qfn97460

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/SigningRSAESPKCSCallbackTests.cs [hunk 5]
```diff
@@ -1,5 +1,5 @@
-/*
- * Copyright 2019-present MongoDB, Inc.
+﻿/*
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/SigningRSAESPKCSCallbackTests.cs:1]: I like this diff with java bindings :)

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/SigningRSAESPKCSCallbackTests.cs [hunk 6]
```diff
@@ -52,10 +40,26 @@ public void testGetSignature() throws InvalidKeySpecException, NoSuchAlgorithmEx
             + "9QP1V3SRW0XoD7ez8FpFabp42cmPOxUNk3FK3paQZABLxH5pzCWI9PzIAVfPDrm+sdnbgG7vAnwfL2IMMJSA3aDYGCbF9EgefG+STcpfqq7fQ6f5TBgLFwKBgCd7"
             + "gn1xYL696SaKVSm7VngpXlczHVEpz3kStWR5gfzriPBxXgMVcWmcbajRser7ARpCEfbxM1UJyv6oAYZWVSNErNzNVb4POqLYcCNySuC6xKhs9FrEQnyKjyk8wI4V"
             + "nrEMGrQ8e+qYSwYk9Gh6dKGoRMAPYVXQAO0fIsHF/T0a";
-    private static final String DATA_TO_SIGN = "data to sign";
-    private static final String EXPECTED_SIGNATURE = "VocBRhpMmQ2XCzVehWSqheQLnU889gf3dhU4AnVnQTJjsKx/CM23qKDPkZDd2A/BnQsp99SN7ksIX5Raj0TPw"
+
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/SigningRSAESPKCSCallbackTests.cs:42]: test data https://docs.google.com/document/d/1-qmr6miUug2qciOlX2Q7eisChvG_RtzLNJnp6ZtmtvY/edit?ts=5f93353d#heading=h.mp30qfn97460

File: bindings/cs/MongoDB.Libmongocrypt.Test/MongoDB.Libmongocrypt.Test.csproj [hunk 7]
```diff
@@ -1,8 +1,8 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>net452;netcoreapp2.1</TargetFrameworks>
-    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp2.1;netcoreapp2.2</TargetFrameworks>
+    <TargetFrameworks>net452;netcoreapp1.1;netcoreapp2.1;netcoreapp3.0</TargetFrameworks>
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt.Test/MongoDB.Libmongocrypt.Test.csproj:4]: as it was discussed now we support 4 TFs for windows and 3 for other systems

File: bindings/cs/MongoDB.Libmongocrypt/Callbacks/CipherCallbacks.cs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                var inputBinary = new Binary(BinarySafeHandle.FromIntPtr(@in));
+                var outputBinary = new Binary(BinarySafeHandle.FromIntPtr(@out));
+                var ivBinary = new Binary(BinarySafeHandle.FromIntPtr(iv));
+
+                byte[] keyBytes = keyBinary.ToArray();
+                byte[] ivBytes = ivBinary.ToArray();
+                byte[] inputBytes = inputBinary.ToArray();
+
+                var outputBytes = AesCrypt(keyBytes, ivBytes, inputBytes, CryptMode.Encrypt);
+                bytes_written = (uint)outputBytes.Length;
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt/Callbacks/CipherCallbacks.cs]: @kevinAlbs why you cannot get the size of `bytes_written` from the `outputBytes`. Why do we need to specify this value on the level of the binding?

**Comment by DmitryLukyanov** [comment]
I will update working with Notsupported exceptions tomorrow

File: bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs [hunk 9]
```diff
@@ -23,14 +24,47 @@ namespace MongoDB.Libmongocrypt
...
         public static CryptClient Create(CryptOptions options)
         {
-            MongoCryptSafeHandle handle = Library.mongocrypt_new();
+            var handle = Library.mongocrypt_new();
+
+            var status = new Status();
+
+            // The below code can be avoided on Windows. So, we don't call it on this system 
+            // to avoid restrictions on target frameworks that present in some of below
+            //if (OperatingSystemHelper.CurrentOperatingSystem != OperatingSystemPlatform.Windows)
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs]: this will be restored. Need to be able making testing on windows as well

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+using System;
+using System.Linq;
+
+namespace MongoDB.Libmongocrypt.Test.Callbacks
+{
+    internal static class CallbackUtils
+    {
+        public static byte[] GetBytesFromHex(string hex)
+        {
+            return Enumerable.Range(0, hex.Length)
```
Comment (markbenvenuto) [inline bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs]: There are more efficient solutions on this page https://stackoverflow.com/questions/311165/how-do-you-convert-a-byte-array-to-a-hexadecimal-string-and-vice-versa

then https://stackoverflow.com/questions/321370/how-can-i-convert-a-hex-string-to-a-byte-array

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/SigningRSAESPKCSCallbackTests.cs (hunk 5 above)
Comment (markbenvenuto) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/SigningRSAESPKCSCallbackTests.cs:1]: :rofl:

File: bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs [hunk 11]
```diff
@@ -23,14 +24,47 @@ namespace MongoDB.Libmongocrypt
...
         /// <param name="options">The options.</param>
         /// <returns>A CryptClient</returns>
         public static CryptClient Create(CryptOptions options)
         {
-            MongoCryptSafeHandle handle = Library.mongocrypt_new();
+            var handle = Library.mongocrypt_new();
+
+            var status = new Status();
+
+            // The below code can be avoided on Windows. So, we don't call it on this system
```
Comment (markbenvenuto) [inline bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs:45]: This is good to know that the native crypto in libmongocrypt still provides value

**Review by markbenvenuto — APPROVED** [review]

File: bindings/cs/MongoDB.Libmongocrypt/Binary.cs [hunk 12]
```diff
@@ -77,14 +77,16 @@ public byte[] ToArray()
...
+            if (Length >= bytes.Length)
             {
-                Marshal.Copy(bytes, 0, Data, (int)Length);
+                Marshal.Copy(bytes, 0, Data, bytes.Length);
             }
             else
             {
                 // this code path is not expected, but it's worth doing it to avoid silent saving of corrupted data
-                throw new InvalidDataException($"Incorrect bytes size. The bytes size must be {Length}.");
+                throw new InvalidDataException($"Incorrect bytes size {bytes.Length}. The bytes size must be {Length}.");
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt/Binary.cs]: I will update this message in the next commit

File: bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs (hunk 9 above)
Comment (jyemin) [reply bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs]: Can you rephrase the second sentence of your comment?  I'm not able to parse it.

File: bindings/cs/MongoDB.Libmongocrypt/Callbacks/CipherCallbacks.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/Callbacks/CipherCallbacks.cs]: is it because the allocated memory for `out` can be bigger than `outputBytes` and you cannot determine(for some reasons) the exact arrays size on the `c` level?

Comment (kevinAlbs) [reply bindings/cs/MongoDB.Libmongocrypt/Callbacks/CipherCallbacks.cs]: It is a little redundant. The expected number of bytes written could be determined in libmongocrypt. But currently it is used as an extra safety check, to ensure that the bindings write an output that is divisible by the block size: https://github.com/mongodb/libmongocrypt/blob/752fc0b6d907561c51e2671ad1bef4936d92d724/src/mongocrypt-crypto.c#L343

File: bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs (hunk 9 above)
Comment (jyemin) [reply bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs]: Sorry, I meant the second sentence of the PR comment!  I don't know what "Need to be able making testing on windows as well" means.  What do you have to do before uncommenting the line of code?

Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs]: > What do you have to do before uncommenting the line of code?

nothing, but if we uncomment this line, the callbacks won't be called (since I test only on windows for now) at all. So I made this temporal change just to show how it can be checked on windows if someone wants to try it

**Review by kevinAlbs — APPROVED** [review]

**Comment by DmitryLukyanov** [comment]
EG: https://evergreen.mongodb.com/version/5faef31e3e8e8655c85049af

File: bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs [hunk 13]
```diff
@@ -23,14 +23,52 @@ namespace MongoDB.Libmongocrypt
...
+            var handle = Library.mongocrypt_new();
+
+            var status = new Status();
 
-            Status status = new Status();
+            // The below code can be avoided on Windows. So, we don't call it on this system 
+            // to avoid restrictions on target frameworks that present in some of below
+            if (OperatingSystemHelper.CurrentOperatingSystem != OperatingSystemPlatform.Windows)
+            {
+#if !NETSTANDARD1_5
```
Comment (DmitryLukyanov) [inline bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs:49]: these `ifdef` conditions are a bit tricky. The reason why we added them is that we want to provide the ability for users to manually provide the `mongocrypt` library which will include a linked OpenSSL library which can be used instead of the added hooks for target frameworks which we don't support. So, we need to avoid exceptions in this case.
This behavior should be tested more deeply on Linux. So, I will test it in the next ticket

**Review by markbenvenuto — APPROVED** [review]

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+using System.Linq;
+
+namespace MongoDB.Libmongocrypt.Test.Callbacks
+{
+    internal static class CallbackUtils
+    {
+        public static byte[] GetBytesFromHex(string hex)
+        {
+            int length = hex.Length;
+            byte[] bytes = new byte[length / 2];
```
Comment (rstam) [inline bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs:31]: What happens if length is odd?

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+            var keyHex = "92faa793d717675e2be804584a8a98252083fe6bf16010546a92e2ef4bdd27fd";
+            var ivHex = "31164b2f661e41fed5df60bfcfa40baa";
+            var inputHex = "379ddb78c30e5e4bf19dd81ae705796f";
+            var expectedHex = "671db60d464b09e9c3b03242dd29bdc5";
+
+            var keyBytes = CallbackUtils.GetBytesFromHex(keyHex);
+            var ivBytes = CallbackUtils.GetBytesFromHex(ivHex);
+            var inputBytes = CallbackUtils.GetBytesFromHex(inputHex); // decryptedBytes
+            var expectedEncryptedBytes = CallbackUtils.GetBytesFromHex(expectedHex);
+#if !NETCOREAPP1_1
```
Comment (rstam) [inline bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs:36]: As a rule I dislike negative tests when there is also an else part.

Although I see the appeal of having the "normal" part in the then part and the "abnormal" part in the else part...

File: bindings/cs/MongoDB.Libmongocrypt.Test/MongoDB.Libmongocrypt.Test.csproj (hunk 7 above)
Comment (rstam) [reply bindings/cs/MongoDB.Libmongocrypt.Test/MongoDB.Libmongocrypt.Test.csproj:4]: Why? Is it because `net452` is Windows only?

What's the point of testing with both `netcoreapp2.1` and `netcoreapp3.0`?

File: bindings/cs/MongoDB.Libmongocrypt/MongoDB.Libmongocrypt.csproj [hunk 16]
```diff
@@ -1,8 +1,8 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>net452;netstandard1.5</TargetFrameworks>
-    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netstandard1.5</TargetFrameworks>
+    <TargetFrameworks>net452;netstandard1.5;netstandard2.0;netstandard2.1</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netstandard1.5;netstandard2.0;netstandard2.1</TargetFrameworks>
```
Comment (rstam) [inline bindings/cs/MongoDB.Libmongocrypt/MongoDB.Libmongocrypt.csproj:5]: How does this work? Is line 5 simply *overwriting* what line 4 did when the `$OS` is not `Windows_NT`?

Comment (rstam) [reply bindings/cs/MongoDB.Libmongocrypt/MongoDB.Libmongocrypt.csproj:5]: BUT... this means we HAVE to build the Nuget package on Windows right?

Because the Nuget package MUST have artifacts for `net452`

File: bindings/cs/MongoDB.Libmongocrypt/Binary.cs [hunk 17]
```diff
@@ -77,14 +77,16 @@ public byte[] ToArray()
         /// </summary>
         public void WriteBytes(byte[] bytes)
         {
-            if (Length == bytes.Length)
+            // The length of the new bytes can be smaller than allocated memory 
+            // because sometimes the allocated memory contains reserved blocks for future usage
+            if (Length >= bytes.Length)
```
Comment (rstam) [inline bindings/cs/MongoDB.Libmongocrypt/Binary.cs]: I think we're trying to determine if it's safe to copy the bytes? The test makes more sense the other way around:

    if (bytes.Length <= Length)

**Review by rstam — COMMENTED** [review]
I've done a light review and have only superficial comments.

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs (hunk 14 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs:31]: Exception, but actually all other ways that I saw (to make this work) had this problem too. It allowed me to find bugs in the test data which is good, because otherwise I could think that it's a problem in the code :) 

I think since it's a test method, it's ok to have it as is

File: bindings/cs/MongoDB.Libmongocrypt.Test/MongoDB.Libmongocrypt.Test.csproj (hunk 7 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt.Test/MongoDB.Libmongocrypt.Test.csproj:4]: >Why? Is it because net452 is Windows only?

It's actually how it was done before. I just updated the list of frameworks. As I mentioned in the slack, I would check this more deeply in the scope of MONGOCRYPT-86

Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt.Test/MongoDB.Libmongocrypt.Test.csproj:4]: >What's the point of testing with both netcoreapp2.1 and netcoreapp3.0?

`netcoreapp2.1` is `netstandard2.0` and `netcoreapp3.0` is `netstandard2.1`. We have a case which works only with `netstandard2.1`.

File: bindings/cs/MongoDB.Libmongocrypt/Binary.cs (hunk 17 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/Binary.cs]: done

File: bindings/cs/MongoDB.Libmongocrypt/MongoDB.Libmongocrypt.csproj (hunk 16 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/MongoDB.Libmongocrypt.csproj:5]: good point, I would leave the current change in the scope of the current ticket since now I just mimic the previous logic which was added by Vincent and will update it in the scope of my current ticket which is exactly about Linux related things including packaging

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs (hunk 15 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs:36]: The second point you mentioned looks more significant for me in this particular case. I would leave it as is

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs (hunk 14 above)
Comment (JamesKovacs) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs:31]: Even though it is test code, I would suggest adding a guard clause such as `hex.Length % 2 == 0` to ensure that the incoming string has the correct number of hex digits.

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs (hunk 15 above)
Comment (JamesKovacs) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs:36]: We are testing two completely different things here. e.g. That encryption succeeds on supported platforms and throws on unsupported ones. I would separate these out into two separate tests and mark them as conditional on the appropriate frameworks.

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/HashCallbackTests.cs (hunk 3 above)
Comment (JamesKovacs) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/HashCallbackTests.cs:26]: Same as before. Testing two different things. We should have two different tests. I won't annoy you by pointing it out for each one but we should separate out the NotSupported tests from the correct impl tests.

File: bindings/cs/MongoDB.Libmongocrypt/HashCallback.cs [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+{
+    internal static class HashCallback
+    {
+        public static bool Hash(
+            IntPtr ctx,
+            IntPtr @in,
+            IntPtr @out,
+            IntPtr statusPtr)
+        {
+            var status = new Status(StatusSafeHandle.FromIntPtr(statusPtr));
```
Comment (JamesKovacs) [inline bindings/cs/MongoDB.Libmongocrypt/HashCallback.cs]: `StatusSafeHandle` derives from `SafeHandle`, which is both `IDisposable` and has a finalizer. If we don't dispose of the `SafeHandle` explicitly, then we force the runtime to go down the finalizer route to clean up the associated resources. We should put this in a `using` block.

File: bindings/cs/MongoDB.Libmongocrypt/HmacShaCallbacks.cs [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        private static bool Hmac(
+            IntPtr key,
+            IntPtr @in,
+            IntPtr @out,
+            IntPtr statusPtr,
+            int bitness)
+        {
+            var status = new Status(StatusSafeHandle.FromIntPtr(statusPtr));
```
Comment (JamesKovacs) [inline bindings/cs/MongoDB.Libmongocrypt/HmacShaCallbacks.cs]: Another `SafeHandle`, which should be encapsulated in a `using` block.

File: bindings/cs/MongoDB.Libmongocrypt/SecureRandomCallback.cs [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    internal static class SecureRandomCallback
+    {
+        public static bool GenerateRandom(
+            IntPtr ctx,
+            IntPtr @out,
+            uint count,
+            IntPtr statusPtr)
+        {
+            var outBinary = new Binary(BinarySafeHandle.FromIntPtr(@out));
+            var status = new Status(StatusSafeHandle.FromIntPtr(statusPtr));
```
Comment (JamesKovacs) [inline bindings/cs/MongoDB.Libmongocrypt/SecureRandomCallback.cs]: `using` blocks `SafeHandle` instances here too.

File: bindings/cs/MongoDB.Libmongocrypt/SigningRSAESPKCSCallback.cs [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+    internal static class SigningRSAESPKCSCallback
+    {
+        public static bool RsaSign(
+            IntPtr ctx,
+            IntPtr key,
+            IntPtr inData,
+            IntPtr outData,
+            IntPtr statusPtr)
+        {
+            var status = new Status(StatusSafeHandle.FromIntPtr(statusPtr));
```
Comment (JamesKovacs) [inline bindings/cs/MongoDB.Libmongocrypt/SigningRSAESPKCSCallback.cs]: `using` around `SafeHandle`.

File: bindings/cs/MongoDB.Libmongocrypt/SigningRSAESPKCSCallback.cs [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+                return false;
+            }
+
+            return true;
+        }
+
+        public static byte[] HashAndSignBytes(byte[] dataToSign, byte[] key)
+        {
+#if NETSTANDARD2_1
+            var rsaProvider = new RSACryptoServiceProvider();
```
Comment (JamesKovacs) [inline bindings/cs/MongoDB.Libmongocrypt/SigningRSAESPKCSCallback.cs]: `RSACryptoServiceProvider` is `IDisposable`. `using` once again.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
`SafeHandle` instances are not being disposed in a number of cases and we are letting them land on the finalizer queue. The other comments are more stylistic in separating out the tests for NotSupported versus testing actual functionality.

File: bindings/cs/MongoDB.Libmongocrypt/HashCallback.cs (hunk 18 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/HashCallback.cs]: good catch, done

File: bindings/cs/MongoDB.Libmongocrypt/HmacShaCallbacks.cs (hunk 19 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/HmacShaCallbacks.cs]: done

File: bindings/cs/MongoDB.Libmongocrypt/SecureRandomCallback.cs (hunk 20 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/SecureRandomCallback.cs]: done

File: bindings/cs/MongoDB.Libmongocrypt/SigningRSAESPKCSCallback.cs (hunk 21 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/SigningRSAESPKCSCallback.cs]: done

File: bindings/cs/MongoDB.Libmongocrypt/SigningRSAESPKCSCallback.cs (hunk 22 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt/SigningRSAESPKCSCallback.cs]: done

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs (hunk 14 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CallbackUtils.cs:31]: done

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs (hunk 15 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs:36]: I think this is fine for this particular test. We test a general use case (not an edge case, not an additional scenario, not boundary values, not negative scenario and etc) how this code can be used. And this general use case is not the same for different frameworks that's why we have this diff in asserting. I agree that it can be split in some meaning, but I don't think that it's reasonable here

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/HashCallbackTests.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/HashCallbackTests.cs:26]: Answered above

**Review by DmitryLukyanov — COMMENTED** [review]
https://evergreen.mongodb.com/version/5fb56ff50ae606292184555e

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs (hunk 15 above)
Comment (rstam) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs:36]: While it is true that the expected behavior is different on some target frameworks I don't object to having this be in one test. The Arrange/Act parts of the test are the same on all platforms. Only the Assert part is different. It makes sense to share the common parts to avoid code duplication.

**Review by rstam — COMMENTED** [review]
I've resolved my comments.

I'm going to let others LGTM because I'm only reviewing superficially.

File: bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs (hunk 15 above)
Comment (JamesKovacs) [reply bindings/cs/MongoDB.Libmongocrypt.Test/Callbacks/CipherCallbacksTests.cs:36]: I'm fine with it as-is then.

File: bindings/cs/MongoDB.Libmongocrypt/CipherCallbacks.cs [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+        public static bool Encrypt(
+            IntPtr ctx,
+            IntPtr key,
+            IntPtr iv,
+            IntPtr @in,
+            IntPtr @out,
+            ref uint bytes_written,
+            IntPtr statusPtr)
+        {
+            var status = new Status(StatusSafeHandle.FromIntPtr(statusPtr));
```
Comment (JamesKovacs) [inline bindings/cs/MongoDB.Libmongocrypt/CipherCallbacks.cs]: Missed a needed `using` block.

File: bindings/cs/MongoDB.Libmongocrypt/CipherCallbacks.cs [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+        public static bool Decrypt(
+            IntPtr ctx,
+            IntPtr key,
+            IntPtr iv,
+            IntPtr @in,
+            IntPtr @out,
+            ref uint bytes_written,
+            IntPtr statusPtr)
+        {
+            var status = new Status(StatusSafeHandle.FromIntPtr(statusPtr));
```
Comment (JamesKovacs) [inline bindings/cs/MongoDB.Libmongocrypt/CipherCallbacks.cs]: Missed a needed `using` block.

File: bindings/cs/MongoDB.Libmongocrypt/SecureRandomCallback.cs [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+{
+    internal static class SecureRandomCallback
+    {
+        public static bool GenerateRandom(
+            IntPtr ctx,
+            IntPtr @out,
+            uint count,
+            IntPtr statusPtr)
+        {
+            var outBinary = new Binary(BinarySafeHandle.FromIntPtr(@out));
```
Comment (JamesKovacs) [inline bindings/cs/MongoDB.Libmongocrypt/SecureRandomCallback.cs]: BinarySafeHandle probably needs a `using` block too.

**Review by JamesKovacs — APPROVED** [review]
LGTM
